### ISSUE TITLE
Unexpected behavior from Layout/CommentIndentation

### ISSUE THREAD

**Issue by grantbdev** [rubocop-hq/rubocop#5247]
`Layout/CommentIndendation` enforces unnatural comment indentation if the next line of code is an outdented `private`.

I have not looked at the cop source code yet, but I suspect that the cop checks the indentation of the next line of code for guidance and thinks that the comment is "for" `private` and thus should align properly. Perhaps a blank line between the comment and `private` could be considered indication that the comment is not necessarily intended to align with the outdented `private`.

--------

## Expected behavior

With the following RuboCop configuration:
```yaml
Layout/AccessModifierIndentation:
  EnforcedStyle: outdent
```

Comments should not expected to be on the same indentation level as the following `private` access modifier.

```ruby
class Foo
  # Any comment, e.g. rubocop:disable Metrics/ParameterLists
  def foo; end
  # Any comment, e.g. rubocop:enable Metrics/ParameterLists

private

  def bar; end
end
```

For the example above, I would expect `no offenses detected`.

## Actual behavior

If the next line of code after a comment line is the `private` access modifier, `Layout/CommentIndentation` expects the comment to have same indentation as the `private` access modifier which is undesirable if `private` is outdented.

For the example above, I get `[Corrected] Layout/CommentIndentation: Incorrect indentation detected (column 2 instead of 0).`

Note that the problem goes away if the next line of code is on the expected indentation level like another method definition. 
Also note that the cop doesn't try to match an outdented `end`.

## Steps to reproduce the problem

With the following RuboCop configuration:
```yaml
Layout/AccessModifierIndentation:
  EnforcedStyle: outdent
```

Inspect the following code:
```ruby
# Example class
class Foo
  # Any comment

private

  def bar; end
end
```

## RuboCop version

0.49.1 (using Parser 2.4.0.0, running on ruby 2.3.5 x86_64-darwin16)

(I tried 0.52 and the behavior still persisted)

**Comment by grantbdev** [comment]
I have attempted a solution here: https://github.com/grantbdev/rubocop/commits/fix_comment_indentation_before_private

There is a straightforward regex solution and an uglier but conceptually more DRY solution that checks the AST node. If one of these is acceptable, I will open a PR.

**Comment by dkubb** [comment]
> There is a straightforward regex solution and an uglier but conceptually more DRY solution that checks the AST node.
>
> The node solution does not look very good, but I thought it was worth showing for consideration.

I'm not a rubocop dev, but the AST version of the code is vastly superior IMHO. The knowledge about what constitutes an access modifier is already encoded in the system so it makes more sense to just call the method to test the type of node. The regexp solution may be expedient, but it's not as resilient to change -- if new access modifiers are added in the future it's very likely this code won't be updated, which will be the source of a bug.

**Comment by bbatsov** [comment]
Indeed. At any rate - solving this is going to be tricky, as we'll have to add special handling for a handful of macro methods. That's one of the reasons why I think that this outdenting is a bad idea in general. We'd happily take a PR that solves this without increasing the code complexity, but I don't think our Core Team will tackle this any time soon.
